File: cracking_code_book/one_away/one_away.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>

bool one_away(std::string, std::string);

int main(int argc, char ** argv){
	std::vector< std::pair < std::string, std::string > > v;
	v.push_back(std::pair<std::string, std::string>("pale","ple"));
	v.push_back(std::pair<std::string, std::string>("zale","zales"));
	v.push_back(std::pair<std::string, std::string>("pale","bale"));
	v.push_back(std::pair<std::string, std::string>("pale","bake"));
	for(auto i : v){
		std::cout << one_away(i.first, i.second) << std::endl;
	}
	return 0;
}

bool one_away(std::string input_string, std::string comparison_string){
	if(abs(input_string.size() - comparison_string.size()) > 1)
		return false;

	short loffset = 0;
	short roffset = 0;
	bool incorrect = false;
	for(int i = 0; i < std::min(input_string.size(), comparison_string.size()); i++) {
		if(input_string[i + loffset] != comparison_string[i + roffset]){
			if(input_string.size() < comparison_string.size())
				loffset--;
			else if(input_string.size() > comparison_string.size())
				roffset--;

			if(incorrect)
				return false;
			else
				incorrect = true;
		}
		if((incorrect == true) && (input_string.back() != comparison_string.back()))
			return false;
	}
	return true;
}

File: cracking_code_book/one_away/one_away.py
tests = [("pale", "ple"), ("zale","zales"), ("pale", "bale"), ("pale","bake")]

def one_away(input_string, comparison_string):
	if abs(len(input_string) - len(comparison_string)) > 1:
		return False

	loffset = 0
	roffset = 0
	incorrect = False
	for i in range(min(len(input_string), len(comparison_string))):
		if input_string[i + loffset] != comparison_string[i + roffset]:
			if len(input_string) < len(comparison_string):
				loffset -= 1
			elif len(input_string) > len(comparison_string):
				roffset -= 1
			if incorrect:
				return False
			else:
				incorrect = True
		if incorrect and input_string[-1] != comparison_string[-1]:
			return False
	return True




for test in tests:
	print one_away(test[0], test[1])